network/tools/network-policy-enforcement-latency/pkg/client.go

```diff
@@ -0,0 +1,552 @@
+			klog.Infof("Test client got pod %q with assigned IP %q, %v after pod creation", podName, pod.Status.PodIP, podAssignedIPLatency)
+		}
+	}
+}
+
+// recordFirstSuccessfulRequest sends curl requests continuously to the provided
+// target IP and target port from the test client config, at 1 second intervals,
+// until the first successful response. It records the latency between the
+// provided start time in the targetSpec, and the time of the first successful
+// response.
+func (c *TestClient) recordFirstSuccessfulRequest(target *targetSpec, forPodCreation bool) {
+	if target == nil || len(target.ip) == 0 {
+		klog.Infof("Skipping a target for policy enforcement latency (podCreation=%t), because the target does not have an IP address assigned. Target: %v", forPodCreation, target)
+		return
+	}
+
+	request := fmt.Sprintf("curl --connect-timeout 1 %s:%d", target.ip, c.config.targetPort)
+	command := []string{"sh", "-c", "-x", request}
+	klog.Infof("Sending requests %q", request)
+
+	for {
+		select {
+		case <-c.MainStopChan:
+			return
+		default:
+		}
+
+		_, err := utils.RunCommand(command)
+		// CURL command returns errors for requests denied by a policy, or when the
+		// IP is unreachable. Everything else is considered a success.
+		if err != nil {
+			continue
+		}
+
+		if forPodCreation {
+			c.reportReachedTimeForPodCreation(target, time.Now())
+		} else {
+			c.reportReachedTimeForPolicyCreation(target, time.Now())
+		}
+		return
+	}
+}
+
+// reportReachedTimeForPodCreation records network policy enforcement latency
+// metric for pod creation.
+func (c *TestClient) reportReachedTimeForPodCreation(target *targetSpec, reachedTime time.Time) {
+	latency := reachedTime.Sub(target.startTime)
+	metrics.PolicyEnforceLatencyPodCreation.Observe(latency.Seconds())
+	klog.Infof("Pod %q in namespace %q reached %v after pod IP was assigned", target.name, target.namespace, latency)
+}
+
+// reportReachedTimeForPolicyCreation records network policy enforcement latency
+// metric for network policy creation.
+func (c *TestClient) reportReachedTimeForPolicyCreation(target *targetSpec, reachedTime time.Time) {
+	if len(c.config.allowPolicyName) == 0 {
+		return
+	}
+
+	var policyCreateTime time.Time
+	failed := false
+
+	c.policyCreatedTime.lock.Lock()
+
+	if c.policyCreatedTime.time == nil {
+		networkPolicy, err := c.k8sClient.NetworkingV1().NetworkPolicies(c.config.testClientNamespace).Get(context.TODO(), c.config.allowPolicyName, metav1.GetOptions{})
+		if err != nil {
+			klog.Warningf("Failed to get network policies for pod %q in namespace %q with IP %q: %v", target.name, c.config.testClientNamespace, target.ip, err)
+			failed = true
+		} else {
+			policyCreateTime = networkPolicy.GetCreationTimestamp().Time
+			c.policyCreatedTime.time = &policyCreateTime
+		}
+	} else {
+		policyCreateTime = *c.policyCreatedTime.time
+	}
+
+	c.policyCreatedTime.lock.Unlock()
+	if failed {
```

Comment:
What happens in this case?
Don't we lose some test data because of this?
Doesn't it mean that request was sent before the policy was deployed?
Can't we wait for the policy before testing reachability?

Reply:
Policy creation test starts sending traffic before allow policy is created.
Once allow policy is created, the requests to pods will succeed, and the policy will exist to measure latency for its creation.
If the policy cannot be retrieved here, it means the setup is incorrect. The warning log will make it clear that the specified allow policy is not found. No metrics should be generated from it.